examples/defun.kl

```diff
@@ -19,12 +26,14 @@
 
 [example (const 'a 'b)]
 
+-- (last-stx foo bar baz) => baz
 [defun last-stx (stx)
   (syntax-case stx
     [[cons a d]
      (syntax-case d
        [() a]
-       [_ (last-stx d)])]
+       [_ (last-stx d)])]  -- note that this is a recursive call to last-stx, whereas earlier (e.g. in n-ary-app.kl)
+                           -- we'd have to return the syntax of a macro call to last-stx.
```

Comment:
Speaking of which, how come that trick works in `n-ary-app.kl`? In the same manner we can write

    (let ((x 2))
      (let ((x (+ x x)))
        x))

And the `x` in `(+ x x)` refers to the outer `x`, I would expect to be able to write something like

    (define-macro (my-macro) #'2)
    (define-macro (my-macro) #'(+ (my-macro) (my-macro)))

and to have the `my-macro` in `#'(+ (my-macro) (my-macro))` refer to the first definition of `my-macro` (unless `define-macro` is like a `letrec`, but I don't think that's the case otherwise we wouldn't need `defun`). In the case of `my-app`, since there is no previous definition of `my-app`, that should be an undefined identifier, I don't think the body of `my-app` should be able to refer to the not-yet-in-scope `my-app`.

Am I understanding this all wrong? Or did we perhaps accidentally add a bug in our scope-set implementation somewhere in order to make the `n-ary-app.kl` test pass when it shouldn't? If so, I think we should just make `define-macro` behave like `letrec`, fix the bug, and skip `defun`.

Reply:
The key here is the way that macro scoping interacts with program scoping.

Before I added the recursive let form on the type checker branch, there were no ordinary recursive definitions in Klister. This was for the sake of minimality, because a Y-combinator is easy to write and stick behind a macro. But part of the point of typed lambda calculi is to rule out Y, so the type system required that addition (also to make the H-M let-generalization do the right thing).

Recursive macro definitions are different from recursive value definitions. A recursive value is defined in terms of itself - its definition refers to its own name. This reference takes the form of a use, and its meaning is given by substitution. A recursive macro is merely able to *mention* itself, and it can emit quoted code that refers to the same macro. But it's not recursive in the sense that a recursive value is.

Reply:
Would it be correct to say that `define-macros` is similar to `letrec` in that it makes the identifier available within the scope of the definition, except that `letrec` makes the identifier available in the current phase whereas `define-macros` only makes it available in the next phase?

Reply:
I don't think so, because the meaning of "making it available" is quite different.  A `letrec` introduces a recursive binding (recursion at core language level), while `define-macros` enables quotations in its body to refer to the macro's name (recursion at user syntax level). So there's a use-mention distinction going on, but also a conceptual difference.